### ISSUE TITLE
Not all eslint 1.0.0 formatters work in grunt

### ISSUE THREAD

**Issue by pkosenko** [eslint/eslint#3437]
I notice that not all the formatters that ship with eslint 1.0.0 actually work with Grunt, although they do work on the eslint command line.  I moved the folder over to another location accessible by grunt and ran the following (running node.js on Windows 7):

    // Using gruntify-eslint
    eslint: {
      options: {
        format: "./formatters/compact",
       },

I replaced "compact" with the filenames (minus .js extension) and got 
the following:

checkstyle	// works
compact		// works
jslint-xml	// Warning: Formatter ./formatters/jslint-xml not found Use --force to continue.
json		// works
junit		// Warning: Formatter ./formatters/junit not found Use --force to continue.
stylish		// Warning: Formatter ./formatters/stylish not found Use --force to continue.
tap		// Warning: Formatter ./formatters/tap not found Use --force to continue.
unix		// works

The failure of "stylish" is strange, since it is the DEFAULT, and seems to work when it is accessed as the default (with no explicit object reference).

I notice that the function processing occurs here is cli.js:

    formatter = engine.getFormatter(format);
    if (!formatter) {
        console.error("Could not find formatter '%s'.", format);
        return false;
    }

That apparently passes, since the error message is NOT "Could not find formatter . . .".

So I am guessing that the problem is somewhere in the try block of getFormatter(), since I have the files located in the DEFAULT location ("./formatters/" + format"), and the getFormatter is loading the 
formatter MODULES (all formatters are exported as modules).

getFormatter() is in cli-engine.js:

    getFormatter: function(format) {

        var formatterPath;

        // default is stylish
        format = format || "stylish";

        // only strings are valid formatters
        if (typeof format === "string") {

            // replace \ with / for Windows compatibility
            format = format.replace(/\\/g, "/");

            // if there's a slash, then it's a file
            if (format.indexOf("/") > -1) {
                formatterPath = path.resolve(process.cwd(), format);
            } else {
                formatterPath = "./formatters/" + format;
            }

            try {
                return require(formatterPath);
            } catch (ex) {
                return null;
            }

        } else {
            return null;
        }
    }

*****************

Note that the formatters with grunt problems WORK when run from eslint directly (rather than grunt):

"tap" formatter:

$ eslint -f tap librivox_search_with_views.js
TAP version 13
1..1
not ok 1 - librivox_search_with_views.js
  ---
  message: Expected indentation of 0 characters.
  severity: error
  data:
    line: 3
    column: 0
    ruleId: indent
  messages:
    - message: Trailing spaces not allowed.
      severity: error
      data:
        line: 142
        column: 10
        ruleId: no-trailing-spaces
  ...

"jslint-xml" formatter

$ eslint -f jslint-xml librivox_search_with_views.js
<?xml version="1.0" encoding="utf-8"?><jslint><file name="librivox_search_with_views.js"><issue line="3" char="0" evidence="    var librivoxSearch = angular.module(&apos
;librivoxSearch&apos;, [&apos;ui.router&apos;, &apos;blndspt.ngPerformance&apos;]);" reason="Expected indentation of 0 characters. (indent)" /><issue line="4" char="0" e
vidence="    librivoxSearch.controller(&apos;librivoxSearchController&apos;, function ($scope, $http, searchData, $state, $location) {" reason="Expected indentation of 0
 characters. (indent)" />

"junit" formatter

$ eslint -f junit librivox_search_with_views.js
<?xml version="1.0" encoding="utf-8"?>
<testsuites>
<testsuite package="org.eslint" time="0" tests="110" errors="110" name="librivox_search_with_views.js">
<testcase time="0" name="org.eslint.indent"><failure message="Expected indentation of 0 characters."><![CDATA[line 3, col 0, Error - Expected indentation of 0 characters
. (indent)]]></failure></testcase>
<testcase time="0" name="org.eslint.indent"><failure message="Expected indentation of 0 characters."><![CDATA[line 4, col 0, Error - Expected indentation of 0 characters
. (indent)]]></failure></testcase>

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by gyandeeps** [comment]
What happens when you use it like this:

```js
options: {
    format: "<format name>",
   },
```

`<format name>` = stylish or checkstyle or the other formats

**Comment by pkosenko** [comment]
Yes, I should have tried that.  As a quick test, at least ONE of the problem formatters ("tap") WORKS that way.  It is no doubt being accessed in the "./node_modules/eslint/lib/formatters" location.

So it seems that the problem has to do with trying to load the formatter modules using PATH info for modules that are not located in the standard eslint archive (which might happen if anyone tries creating their own custom formatter).


    // Using gruntify-eslint
    eslint: {
      options: {
        format: "tap"
      }, 

$ grunt eslint:browserFiles
Running "eslint:browserFiles" (eslint) task
TAP version 13
1..4
not ok 1 - Gruntfile.js
  ---
  message: Mixed spaces and tabs.
  severity: error
  data:
    line: 20
    column: 2
    ruleId: no-mixed-spaces-and-tabs
  messages:
    - message: Trailing spaces not allowed.
      severity: error
      data:
        line: 47
        column: 47
        ruleId: no-trailing-spaces
  ...
not ok 2 - librivox_search.js
  ---
  message: 'Use the function form of "use strict".'
  severity: error
..................

**Comment by nzakas** [comment]
This seems like an issue to file with the grunt plugin?

**Comment by gyandeeps** [comment]
> I moved the folder over to another location accessible by grunt

What does this mean?

**Comment by pkosenko** [comment]
Nicholas may be right that it is a grunt or gruntify-eslint issue, but I am quite sure yet.  It didn't seem to me that gruntify-eslint was ITSELF doing any formatter module loading (but I may be mistaken).  That seemed to be done by eslint's engine.getFormatter(format).

I COPIED the "formatters" folder into the top level of the project to see how the PATH reference would work with Grunt there.  Tell me if I have got this wrong (i.e., my misconstruction -- but why did SOME of the paths work and others didn'''t?):

- project folder
     gruntfile.js
     - formatters  (grunt would be accessing formatters in this folder when specified by path?)
       stylish.js
       tap.js
       compact.js
       . . . .
     - node_modules
         - eslint
             - lib
                 - formatters   (eslint command line and is accessing formatters here, and non-path grunt format references)
                   stylish.js
                   tap.js
                   . . . .

**Comment by nzakas** [comment]
@pkosenko I'd suggest filing an issue with the grunt plugin and let them investigate this. The interface you're using is `gruntify-eslint` and that's failing you, so there's likely an issue in the plugin itself that others will encounter as well. If it can be root caused back to us, then we can reopen this issue with more details.

**Comment by pkosenko** [comment]
Actually, the code in gruntify-eslint points to eslint.getFormatters() as the processing function that is handling the reference to the formatters (either module name or path).  So it looks to me like eslint should be looking into that.  It doesn't look like a grunt issue at all.

        formatter = eslint.getFormatter(options.format);
        if (!formatter) {
            grunt.warn("Formatter " + options.format + " not found");
            return;
        }

That is where the error is being generated. eslint.getFormatter() is returning a bad formatter module reference.
